java/core/src/java/org/apache/orc/impl/WriterImpl.java

```diff
@@ -683,20 +683,22 @@ public void addUserMetadata(String name, ByteBuffer value) {
 
   @Override
   public void addRowBatch(VectorizedRowBatch batch) throws IOException {
+    InternalVectorizedRowBatch internalBatch = InternalVectorizedRowBatch.encapsulation(batch);
+    int batchSize = internalBatch.size();
     try {
       // If this is the first set of rows in this stripe, tell the tree writers
       // to prepare the stripe.
-      if (batch.size != 0 && rowsInStripe == 0) {
+      if (batchSize != 0 && rowsInStripe == 0) {
         treeWriter.prepareStripe(stripes.size() + 1);
       }
       if (buildIndex) {
         // Batch the writes up to the rowIndexStride so that we can get the
         // right size indexes.
         int posn = 0;
-        while (posn < batch.size) {
-          int chunkSize = Math.min(batch.size - posn,
+        while (posn < batchSize) {
+          int chunkSize = Math.min(batchSize - posn,
               rowIndexStride - rowsInIndex);
-          treeWriter.writeRootBatch(batch, posn, chunkSize);
+          treeWriter.writeRootBatch(internalBatch, posn, chunkSize);
```

Comment:
If I understand correctly, the selected vector only applies to each row in the root VectorizedRowBatch. We can simply use a bool array in the root VectorizedRowBatch to indicate whether each row is selected or not. Then we can split the input VectorizedRowBatch into smaller pieces of continuously selected VectorizedRowBatchs and thus change posn & chunkSize accordingly.

Reply:
Reusing Selected within a VectorizedRowBatch actually ensures consistent processing across types.
Let's look at a complex case provided in the test case 
```java
  public void testWriteComplexTypeUseSelectedVector() throws IOException {
    TypeDescription schema =
        TypeDescription.fromString("struct<a:map<int,uniontype<int,string>>," +
            "b:array<struct<c:int>>>");
    ....
 }    
```
For example, we have three rows of data, and the third row is selected after the filter. Before reading the third map, we need to determine the offsets of the key and value fields, since the first two maps may have multiple key-value pairs. Of course, if the key and value are also complex types, recursion may be required.

So processing only at the entry does not cover all cases and each type needs to be processed separately.

Imagine that the recursive processing eventually comes to the base type, which doesn't know what offset means. offset may be relative to the row, or it may be a value that has gone through multiple jumps.

Reply:
I think the root treeWriter can deal with selected nested columns very well. Let me demonstrate my idea in this patch: https://github.com/wgtmac/orc/commit/42ee957be4113af62ccb59319d861fb9462a7c1e

Reply:
I feel a sense of clarity,  @wgtmac thank you for the idea, I will use my test cases later to verify

Reply:
The test passed perfectly. Very nice implementation. I'll close the pr. Wait for your pr and I can contribute test cases

Reply:
Thanks for the verification! I will do it later.